src/Compilers/Core/Portable/Diagnostic/SourceLocation.cs

```diff
@@ -22,22 +22,22 @@ public SourceLocation(SyntaxTree syntaxTree, TextSpan span)
         }
 
         public SourceLocation(SyntaxNode node)
-            : this(node.SyntaxTree, node.Span)
+            : this(node.SyntaxTree, node.FullSpan)
```

Comment:
as previously mentioned, this seems entirely wrong.  Many places expect to be able to get the location of something as the expected place to either navigate to or squiggle.  it should not be the FullSpan.

Reply:
I missed that comment before. The other approach I can think of is to change FindNode to not use the FullSpan to compare. That seems problematic as well, and still doesn't guarantee `GetLocation().FindNode()` will round trip correctly.

Reply:
There's an alternate approach you could take.  For example:

You could make it so that Location has SourceSpan and SourceFullSpan.  FindNode can then use the latter.

Reply:
🤔 Then it would better mirror what SyntaxNode has. I like it!

Reply:
Why is it not possible to create a location with a span that the algorithm needs (there is an overload that takes a tree and a span) instead of changing behavior of an existing API or adding information that is going to be unnecessary most of the time?

Reply:
In the case of #29793 it's because the original SyntaxNode is `LocalDeclarationStatementSyntax` and FindNode is finding the contained `VariableDeclarationSyntax`. The only difference is that the `LocalDeclarationStatementSyntax` includes the trailing trivia, which in the repro is an errant trailing `)`.  Without the context of trivia we would never be able to fully recover the original syntax afaik. We need to keep walking up the tree looking for the same `FullSpan` we started with. 

#30366 has some context on this as well

Reply:
> Why is it not possible to create a location with a span that the algorithm needs (there is an overload that takes a tree and a span) instead of changing behavior of an existing API or adding information that is going to be unnecessary most of the time?

@AlekseyTs This pushes the work onto all features that use Locations to roundtrip a node (i.e. a *lot* of analyzers).  Our messaging up to now has been:

> if you want to find a node between analysis and fixing, then use AdditionalLocaitons, and just add `someNode.GetLocation()` to it.  You can then find that later with `someLocation.FindNode()`."

This symmetry is very desirable and powerful.  Unfortunately, it gets complex, fast, when you have spans vs full-spans involved.  `FullSpan`s allow the algorithm to stay super simple and easy to reason about, while also providing a very valuable and easy to understand mechanism for analyzers/fixers, at the cost of a small amount of additional data stored here.